src/docfx/build/Build.cs

```diff
@@ -29,21 +30,36 @@ private static List<Document> GlobFiles(Context context, Docset docset)
                            .ToList();
         }
 
-        private static Task BuildFiles(Context context, List<Document> files)
+        private static Task BuildFiles(Context context, List<Document> files, TableOfContentsMap tocMap)
         {
-            return ParallelUtility.ForEach(files, file => BuildOneFile(context, file));
+            return ParallelUtility.ForEach(files, file => BuildOneFile(context, file, tocMap));
+        }
+
+        private static async Task<TableOfContentsMap> BuildTocMap(Context context, List<Document> files)
+        {
+            var builder = new TableOfContentsMapBuilder();
+
+            var tocFiles = files.Where(f => f.ContentType == ContentType.TableOfContents);
+            if (!tocFiles.Any())
+            {
+                return builder.Build();
+            }
+
+            await ParallelUtility.ForEach(tocFiles, file => BuildTableOfContents.BuildTocMap(context, file, builder));
+
+            return builder.Build();
         }
 
-        private static Task BuildOneFile(Context context, Document file)
+        private static Task BuildOneFile(Context context, Document file, TableOfContentsMap tocMap)
         {
             switch (file.ContentType)
             {
                 case ContentType.Asset:
                     return BuildAsset(context, file);
                 case ContentType.Markdown:
-                    return BuildMarkdown.Build(context, file);
+                    return BuildMarkdown.Build(context, file, tocMap);
```

Comment:
>tocMap [](start = 62, length = 6)

what about move to context?

Reply:
it's important to define the input and output clearly for build pipleline, it's agreement of our design meeting.

---
In reply to: [184597371](https://github.com/dotnet/docfx/pull/2696#discussion_r184597371) [](ancestors = 184597371)

Reply:
Is context part of input?

---
In reply to: [184599311](https://github.com/dotnet/docfx/pull/2696#discussion_r184599311) [](ancestors = 184599311,184597371)